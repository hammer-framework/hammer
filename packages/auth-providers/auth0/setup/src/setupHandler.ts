import fs from 'fs'
import path from 'path'

import { standardAuthHandler } from '@redwoodjs/cli-helpers'

import { Args } from './setup'

const { version } = JSON.parse(
  fs.readFileSync(path.resolve(__dirname, '../package.json'), 'utf-8')
)

export async function handler({ force: forceArg }: Args) {
  standardAuthHandler({
    basedir: __dirname,
    forceArg,
    provider: 'auth0',
    authDecoderImport:
      "import { authDecoder } from '@redwoodjs/auth-auth0-api'",
    apiPackages: [`@redwoodjs/auth-auth0-api@${version}`],
    webPackages: [
      '@auth0/auth0-spa-js@1.22.5',
      `@redwoodjs/auth-auth0-web@${version}`,
    ],
    notes: [
      'You will need to create several environment variables with your Auth0 config options.',
      'Check out web/src/App.{js,tsx} for the variables you need to add.',
      'See: https://auth0.com/docs/quickstart/spa/react#get-your-application-keys',
      '',
      "You must also create an API and set the audience parameter, or you'll",
      'receive an opaque token instead of the required JWT token.',
      'See: https://auth0.com/docs/quickstart/spa/react/02-calling-an-api#create-an-api',
      '',
      'If you want to allow users to get refresh tokens while offline,',
      'you must also enable the Allow Offline Access switch in your',
      'Auth0 API Settings as part of setup configuration.',
      'See: https://auth0.com/docs/tokens/refresh-tokens',
      '',
      'You can increase security by using refresh token rotation which issues a new refresh token',
      'and invalidates the predecessor token with each request made to Auth0 for a new access token.',
      'Rotating the refresh token reduces the risk of a compromised refresh token.',
      'See: https://auth0.com/docs/tokens/refresh-tokens/refresh-token-rotation',
    ],
  })
}
