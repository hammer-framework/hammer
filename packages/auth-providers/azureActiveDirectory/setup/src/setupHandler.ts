import fs from 'fs'
import path from 'path'

import { standardAuthHandler } from '@redwoodjs/cli-helpers'

import { Args } from './setup'

const { version } = JSON.parse(
  fs.readFileSync(path.resolve(__dirname, '../package.json'), 'utf-8')
)

export async function handler({ force: forceArg }: Args) {
  standardAuthHandler({
    basedir: __dirname,
    forceArg,
    provider: 'azureActiveDirectory',
    authDecoderImport:
      "import { authDecoder } from '@redwoodjs/auth-azure-active-directory-api'",
    apiPackages: [`@redwoodjs/auth-azure-active-directory-api@${version}`],
    webPackages: [
      `@redwoodjs/auth-azure-active-directory-web@${version}`,
      '@azure/msal-browser',
    ],
    notes: [
      'You will need to create several environment variables with your Azure',
      'AD config options. Check out web/src/App.{js,tsx} for the variables',
      'you need to add.',
      '\n',
      'RedwoodJS specific Documentation:',
      'https://redwoodjs.com/docs/authentication#azure-ad',
      '\n',
      'MSAL.js Documentation:',
      'https://docs.microsoft.com/en-us/azure/active-directory/develop/msal-js-initializing-client-applications',
    ],
  })
}
