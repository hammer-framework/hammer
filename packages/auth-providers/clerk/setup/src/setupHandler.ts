import fs from 'fs'
import path from 'path'

import { standardAuthHandler } from '@redwoodjs/cli-helpers'

import { Args } from './setup'

const { version } = JSON.parse(
  fs.readFileSync(path.resolve(__dirname, '../package.json'), 'utf-8')
)

export const handler = async ({ force: forceArg }: Args) => {
  standardAuthHandler({
    basedir: __dirname,
    forceArg,
    authDecoderImport: `import { authDecoder } from '@redwoodjs/auth-clerk-api'`,
    provider: 'clerk',
    webPackages: [
      '@clerk/clerk-react@3.5.1',
      `@redwoodjs/auth-clerk-web@${version}`,
    ],
    apiPackages: [`@redwoodjs/auth-clerk-api@${version}`],
    notes: [
      "You'll need to add three env vars to your .env file:",
      '',
      '```title=".env"',
      'CLERK_FRONTEND_API_URL="Frontend API Key"',
      'CLERK_API_KEY="Backend API Key"',
      'CLERK_JWT_KEY="JWT Verification Key"',
      '```',
      '',
      `You can find their values under "API Keys" on your Clerk app's dashboard.`,
      'Be sure to include `CLERK_FRONTEND_API_URL` in the `includeEnvironmentVariables` array in redwood.toml.',
      '',
      '```toml title="redwood.toml"',
      'includeEnvironmentVariables = [',
      '  "CLERK_FRONTEND_API_URL"',
      '  "CLERK_API_KEY"',
      ']',
      '```',
      '',
      'Also see https://redwoodjs.com/docs/auth/clerk for a full walkthrough.',
    ],
  })
}
