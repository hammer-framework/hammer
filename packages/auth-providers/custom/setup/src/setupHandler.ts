import fs from 'fs'
import path from 'path'

import {
  isTypeScriptProject,
  standardAuthHandler,
} from '@redwoodjs/cli-helpers'

import { Args } from './setup'

const { version } = JSON.parse(
  fs.readFileSync(path.resolve(__dirname, '../package.json'), 'utf-8')
)

export async function handler({ force: forceArg }: Args) {
  const authFilename = isTypeScriptProject() ? 'auth.ts' : 'auth.js'

  standardAuthHandler({
    basedir: __dirname,
    forceArg,
    provider: 'custom',
    webPackages: [`@redwoodjs/auth@${version}`],
    notes: [
      'Done! But you have a little more work to do:\n',
      'You will have to write the actual auth implementation/integration',
      `yourself. Take a look in ${authFilename} and do the necessary changes.`,
    ],
  })
}
