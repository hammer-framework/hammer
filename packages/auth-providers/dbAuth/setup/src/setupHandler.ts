import fs from 'fs'
import path from 'path'

import prompts from 'prompts'

import { getGraphqlPath, standardAuthHandler } from '@redwoodjs/cli-helpers'

import type { Args } from './setup'
import {
  notes,
  notesCreatedUserModel,
  extraTask,
  createUserModelTask,
} from './setupData'
import { hasModel } from './shared'
import {
  notes as webAuthnNotes,
  notesCreatedUserModel as webAuthnNotesCreatedUserModel,
  extraTask as webAuthnExtraTask,
  webPackages as webAuthnWebPackages,
  apiPackages as webAuthnApiPackages,
  createUserModelTask as webAuthnCreateUserModelTask,
} from './webAuthn.setupData'

export async function handler({
  webauthn,
  createUserModel,
  force: forceArg,
}: Args) {
  const { version } = JSON.parse(
    fs.readFileSync(path.resolve(__dirname, '../package.json'), 'utf-8'),
  )

  const webAuthn = await shouldIncludeWebAuthn(webauthn)
  const createDbUserModel = await shouldCreateUserModel(createUserModel)

  let oneMoreThing: string[] = []

  if (webAuthn) {
    if (createDbUserModel) {
      oneMoreThing = webAuthnNotesCreatedUserModel
    } else {
      oneMoreThing = webAuthnNotes
    }
  } else {
    if (createDbUserModel) {
      oneMoreThing = notesCreatedUserModel
    } else {
      oneMoreThing = notes
    }
  }

  let createDbUserModelTask: typeof createUserModelTask | undefined = undefined

  if (createDbUserModel) {
    if (webAuthn) {
      createDbUserModelTask = webAuthnCreateUserModelTask
    } else {
      createDbUserModelTask = createUserModelTask
    }
  }

  standardAuthHandler({
    basedir: __dirname,
    forceArg,
    provider: 'dbAuth',
    authDecoderImport:
      "import { createAuthDecoder } from '@redwoodjs/auth-dbauth-api'",
    webAuthn,
    webPackages: [
      `@redwoodjs/auth-dbauth-web@${version}`,
      ...(webAuthn ? webAuthnWebPackages : []),
    ],
    apiPackages: [
      `@redwoodjs/auth-dbauth-api@${version}`,
      ...(webAuthn ? webAuthnApiPackages : []),
    ],
    extraTasks: [
      webAuthn ? webAuthnExtraTask : extraTask,
      createDbUserModelTask,
      createAuthDecoderFunction,
    ],
    notes: oneMoreThing,
  })
}

/**
 * Prompt the user (unless already specified on the command line) if they want
 * to enable WebAuthn support
 */
async function shouldIncludeWebAuthn(webauthn: boolean | null) {
  if (webauthn === null) {
    const webAuthnResponse = await prompts({
      type: 'confirm',
      name: 'answer',
      message: `Enable WebAuthn support (TouchID/FaceID)? See https://redwoodjs.com/docs/auth/dbAuth#webAuthn`,
      initial: false,
    })

    return webAuthnResponse.answer
  }

  return webauthn
}

/**
 * Prompt the user (unless already specified on the command line) if they want
 * to create a User model in their Prisma schema
 */
async function shouldCreateUserModel(createUserModel: boolean | null) {
  const hasUserModel = await hasModel('User')

  if (createUserModel === null && !hasUserModel) {
    const createModelResponse = await prompts({
      type: 'confirm',
      name: 'answer',
      message: 'Create User model?',
      initial: false,
    })

    return createModelResponse.answer
  }

  return createUserModel
}

export const createAuthDecoderFunction = {
  title: 'Create auth decoder function',
  task: () => {
    const graphqlPath = getGraphqlPath()

    if (!graphqlPath) {
      throw new Error('Could not find your graphql file path')
    }

    const authDecoderCreation =
      'const authDecoder = createAuthDecoder(cookieName)'

    const content = fs.readFileSync(graphqlPath, 'utf-8')

    let newContent = content.replace(
      'import { getCurrentUser } from',
      'import { cookieName, getCurrentUser } from',
    )

    const authDecoderCreationRegexp = new RegExp(
      '^' + escapeRegExp(authDecoderCreation),
      'm',
    )

    if (!authDecoderCreationRegexp.test(newContent)) {
      newContent = newContent.replace(
        'export const handler = createGraphQLHandler({',
        authDecoderCreation +
          '\n\n' +
          'export const handler = createGraphQLHandler({',
      )
    }

    if (!newContent.includes('import { cookieName')) {
      throw new Error('Failed to import cookieName')
    }

    fs.writeFileSync(graphqlPath, newContent)
  },
}

// https://developer.mozilla.org/en-US/docs/Web/JavaScript/Guide/Regular_Expressions#escaping
function escapeRegExp(str: string) {
  return str.replace(/[.*+?^${}()|[\]\\]/g, '\\$&') // $& means the whole matched string
}
