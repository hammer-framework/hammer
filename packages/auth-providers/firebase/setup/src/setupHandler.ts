import fs from 'fs'
import path from 'path'

import { standardAuthHandler } from '@redwoodjs/cli-helpers'

import { Args } from './setup'

const { version } = JSON.parse(
  fs.readFileSync(path.resolve(__dirname, '../package.json'), 'utf-8')
)

export async function handler({ force: forceArg }: Args) {
  standardAuthHandler({
    basedir: __dirname,
    forceArg,
    provider: 'firebase',
    authDecoderImport:
      "import { authDecoder } from '@redwoodjs/auth-firebase-api'",
    webPackages: ['firebase', `@redwoodjs/auth-firebase-web@${version}`],
    apiPackages: ['firebase-admin', `@redwoodjs/auth-firebase-api@${version}`],
    notes: [
      'You will need to create several environment variables with your Firebase config options.',
      'Check out web/src/auth.{js,ts} for the variables you need to add.',
      'See: https://firebase.google.com/docs/web/setup#config-object',
    ],
  })
}
