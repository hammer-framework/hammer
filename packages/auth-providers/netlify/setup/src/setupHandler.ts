import fs from 'fs'
import path from 'path'

import { standardAuthHandler } from '@redwoodjs/cli-helpers'

import { Args } from './setup'

const { version } = JSON.parse(
  fs.readFileSync(path.resolve(__dirname, '../package.json'), 'utf-8')
)

export async function handler({ force: forceArg }: Args) {
  standardAuthHandler({
    basedir: __dirname,
    forceArg,
    provider: 'netlify',
    authDecoderImport: `import { authDecoder } from '@redwoodjs/auth-netlify-api@${version}'`,
    apiPackages: [`@redwoodjs/auth-netlify-api@${version}`],
    webPackages: [
      `@redwoodjs/auth-netlify-web@${version}`,
      'netlify-identity-widget@1.9.2',
    ],
    notes: [
      'You will need to enable Identity on your Netlify site and configure the API endpoint.',
      'See: https://github.com/netlify/netlify-identity-widget#localhost',
    ],
  })
}
