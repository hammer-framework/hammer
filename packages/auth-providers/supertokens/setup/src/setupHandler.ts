import fs from 'fs'
import path from 'path'

import { standardAuthHandler } from '@redwoodjs/cli-helpers'

import { Args } from './setup'

const { version } = JSON.parse(
  fs.readFileSync(path.resolve(__dirname, '../package.json'), 'utf-8')
)

export async function handler({ force: forceArg }: Args) {
  standardAuthHandler({
    basedir: __dirname,
    forceArg,
    provider: 'supertokens',
    authDecoderImport:
      "import { authDecoder } from '@redwoodjs/auth-supertokens-api'",
    apiPackages: [
      `@redwoodjs/auth-supertokens-api@${version}`,
      'supertokens-node',
    ],
    webPackages: [
      `@redwoodjs/auth-supertokens-web@${version}`,
      'supertokens-auth-react',
    ],
    notes: [
      "We've generated some example recipe implementations, but do feel free",
      'to switch to something else that better fit your needs.',
      'See: https://supertokens.com/docs/guides',
    ],
  })
}
