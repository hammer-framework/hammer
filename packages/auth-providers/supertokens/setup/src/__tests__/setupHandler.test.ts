// mock Telemetry for CLI commands so they don't try to spawn a process
jest.mock('@redwoodjs/telemetry', () => {
  return {
    errorTelemetry: () => jest.fn(),
    timedTelemetry: () => jest.fn(),
  }
})

jest.mock('@redwoodjs/cli-helpers', () => {
  return {
    getPaths: () => {
      return {
        web: {
          src: '',
          app: '',
          routes: 'Routes.tsx',
        },
        base: '',
      }
    },
    standardAuthHandler: () => jest.fn(),
  }
})

// This will load packages/auth-providers/supertokens/setup/__mocks__/fs.js
jest.mock('fs')

const mockFS = fs as unknown as Omit<jest.Mocked<typeof fs>, 'readdirSync'> & {
  __setMockFiles: (files: Record<string, string>) => void
}

import fs from 'fs'

import { addRoutingLogic } from '../setupHandler'

describe('addRoutingLogic', () => {
  it('modifies the Routes.{jsx,tsx} file', () => {
    mockFS.__setMockFiles({
      'Routes.tsx':
        "// In this file, all Page components from 'src/pages' are auto-imported.\n" +
        `
import { Router, Route } from '@redwoodjs/router'

import { useAuth } from './auth'

const Routes = () => {
  return (
    <Router useAuth={useAuth}>
      <Route path="/login" page={LoginPage} name="login" />
      <Route path="/signup" page={SignupPage} name="signup" />
      <Route notfound page={NotFoundPage} />
    </Router>
  )
}

export default Routes
`,
    })

    addRoutingLogic.task()

    expect(mockFS.readFileSync('Routes.tsx')).toMatchInlineSnapshot(`
          "// In this file, all Page components from 'src/pages' are auto-imported.

          import { canHandleRoute, getRoutingComponent } from 'supertokens-auth-react/ui'

          import { Router, Route } from '@redwoodjs/router'

          import { useAuth, PreBuiltUI } from './auth'

          const Routes = () => {
            if (canHandleRoute(PreBuiltUI)) {
              return getRoutingComponent(PreBuiltUI)
            }

            return (
              <Router useAuth={useAuth}>
                <Route path="/login" page={LoginPage} name="login" />
                <Route path="/signup" page={SignupPage} name="signup" />
                <Route notfound page={NotFoundPage} />
              </Router>
            )
          }

          export default Routes
          "
      `)
  })

  it('handles a Routes.{jsx,tsx} file with a legacy setup', () => {
    mockFS.__setMockFiles({
      'Routes.tsx':
        "// In this file, all Page components from 'src/pages' are auto-imported.\n" +
        `
import SuperTokens from 'supertokens-auth-react'

import { Router, Route } from '@redwoodjs/router'

import { useAuth } from './auth'

const Routes = () => {
  if (SuperTokens.canHandleRoute()) {
    return SuperTokens.getRoutingComponent()
  }

  return (
    <Router useAuth={useAuth}>
      <Route path="/login" page={LoginPage} name="login" />
      <Route path="/signup" page={SignupPage} name="signup" />
      <Route notfound page={NotFoundPage} />
    </Router>
  )
}

export default Routes
`,
    })

    addRoutingLogic.task()

    expect(mockFS.readFileSync('Routes.tsx')).toMatchInlineSnapshot(`
      "// In this file, all Page components from 'src/pages' are auto-imported.



      import { canHandleRoute, getRoutingComponent } from 'supertokens-auth-react/ui'

      import { Router, Route } from '@redwoodjs/router'

      import { useAuth, PreBuiltUI } from './auth'

      const Routes = () => {
        if (canHandleRoute(PreBuiltUI)) {
          return getRoutingComponent(PreBuiltUI)
        }

        

        return (
          <Router useAuth={useAuth}>
            <Route path="/login" page={LoginPage} name="login" />
            <Route path="/signup" page={SignupPage} name="signup" />
            <Route notfound page={NotFoundPage} />
          </Router>
        )
      }

      export default Routes
      "
    `)
  })
})
