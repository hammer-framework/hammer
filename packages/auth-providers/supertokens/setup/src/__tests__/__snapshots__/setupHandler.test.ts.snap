// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`extraTask 1`] = `
"// In this file, all Page components from 'src/pages' are auto-imported.

import SuperTokens from 'supertokens-auth-react'

import { Router, Route } from '@redwoodjs/router'

import { useAuth } from './auth'

const Routes = () => {
  if (SuperTokens.canHandleRoute()) {
    return SuperTokens.getRoutingComponent()
  }

  return (
    <Router useAuth={useAuth}>
      <Route path="/login" page={LoginPage} name="login" />
      <Route path="/signup" page={SignupPage} name="signup" />
      <Route notfound page={NotFoundPage} />
    </Router>
  )
}

export default Routes
"
`;
